### PR TITLE
Zero-Reference Deep Curve Estimation for low-light image enhancement

### PR DESCRIPTION
**Pull request by soumik12345**
**Zero-Reference Deep Curve Estimation** or **Zero-DCE** formulates light enhancement as a task of image-specific curve estimation with a deep neural network. In this example, we would train a lightweight deep network, **DCE-Net**, to estimate pixel-wise and high-order curves for dynamic range adjustment of a given image. The curve estimation is specially designed, considering pixel value range, monotonicity, and differentiability. Zero-DCE is appealing in its relaxed assumption on reference images, i.e., it does not require any paired or unpaired data during training. This is achieved through a set of carefully formulated non-reference loss functions, which implicitly measure the enhancement quality and drive the learning of the network.

Please find the corresponding Colab notebook [here](https://colab.research.google.com/drive/1EwQ66NiLiqiHVyCLTX7A_LeiY0Jj_QhF?usp=sharing).

### CODE REVIEW

**Comment by soumik12345** [comment]
Hi @fchollet , can you please review this PR?

File: examples/vision/zero_dce.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+Author: [Soumik Rakshit](http://github.com/soumik12345)
+Date created: 2021/09/18
+Last modified: 2021/09/18
+Description: Implementing Zero-Reference Deep Curve Estimation for low-light image enhancement
+"""
+"""
+## Introduction
+
+**Zero-Reference Deep Curve Estimation** or **Zero-DCE** formulates light enhancement
+as a task of image-specific curve estimation with a deep neural network. In this example,
```
Comment (fchollet) [inline examples/vision/zero_dce.py:12]: We need to define curve estimation the first time we mention it, since the reader may not be familiar with it.

File: examples/vision/zero_dce.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+"""
+### Color Constancy Loss
+
+The Color Constancy Loss is used to correct the potential color deviations in the
+enhanced image and also build the relations among the three adjusted channels.
+"""
+
+
+def color_constancy_loss(x):
+    mean_rgb = tf.reduce_mean(x, [1, 2], keepdims=True)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:176]: Pass `axis` as a keyword arg for better readability, and prefer using a tuple

File: examples/vision/zero_dce.py [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+    mr, mg, mb = mean_rgb[:, :, :, 0], mean_rgb[:, :, :, 1], mean_rgb[:, :, :, 2]
+    d_rg = tf.pow(mr - mg, 2)
+    d_rb = tf.pow(mr - mb, 2)
+    d_gb = tf.pow(mb - mg, 2)
+    return tf.pow(tf.pow(d_rg, 2) + tf.pow(d_rb, 2) + tf.pow(d_gb, 2), 0.5)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:181]: `tf.square` is more readable

File: examples/vision/zero_dce.py [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+### Exposure loss
+
+To restrain under-/over-exposed regions, the Exposure Control Loss is used. The exposure
+control loss measures the distance between the average intensity value of a local region
+to the well-exposedness level which is set to `0.6`.
+"""
+
+
+def exposure_loss(x, mean_val=0.6):
+    x = tf.reduce_mean(x, 3, keepdims=True)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:194]: Likewise, prefer keyword args for non-input args

File: examples/vision/zero_dce.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+"""
+### Spatial Consistancy Loss
+
+The spatial consistency loss encourages spatial coherence of the enhanced image through
+preserving the difference of neighboring regions between the input image and its
+enhanced version.
+"""
+
+
+class SpatialConsistancyLoss:
```
Comment (fchollet) [inline examples/vision/zero_dce.py:230]: This loss matches the expected signature `y_true, y_pred` so prefer subclass the `Loss` class

File: examples/vision/zero_dce.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+
+To restrain under-/over-exposed regions, the Exposure Control Loss is used. The exposure
+control loss measures the distance between the average intensity value of a local region
+to the well-exposedness level which is set to `0.6`.
+"""
+
+
+def exposure_loss(x, mean_val=0.6):
+    x = tf.reduce_mean(x, 3, keepdims=True)
+    mean = tf.keras.layers.AveragePooling2D(pool_size=16, strides=16)(x)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:195]: Semantically, you don't need a layer here, so you can use `tf.nn` instead

File: examples/vision/zero_dce.py [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+        x = x + r6 * (tf.pow(x, 2) - x)
+        x = x + r7 * (tf.pow(x, 2) - x)
+        enhanced_image = x + r8 * (tf.pow(x, 2) - x)
+        return enhanced_image
+
+    def call(self, data):
+        dce_net_output = self.dce_model(data)
+        return self.get_enhanced_image(data, dce_net_output)
+
+    def _calculate_losses(self, data, output):
```
Comment (fchollet) [inline examples/vision/zero_dce.py:328]: `compute_losses` would be more in line with other APIs. No need to make it private, either.

In fact this has no dependencies on `self` so it could be better as a standalone utility function rather than a method

File: examples/vision/zero_dce.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+        output = self.dce_model(data)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:329]: This was already computed previously, you already have `output`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

Since we're already published an example on low-light image enhancement, I would generally recommend focusing on new tasks that we haven't covered yet. That said, this is a very nice and high-quality walkthrough, so we should publish it anyway.

File: examples/vision/zero_dce.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+from glob import glob
+from PIL import Image, ImageOps
+import matplotlib.pyplot as plt
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+"""shell
+!gdown https://drive.google.com/uc?id=1DdGIJ4PZPlF2ikl8mNM9V-PdVxVLbQi6
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Removed `!` from shell commnads.

File: examples/vision/zero_dce.py (hunk 1 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Included a basic explanation or curve estimation along with relevant references.

File: examples/vision/zero_dce.py (hunk 3 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Changed to `tf.square` and `tf.sqrt` in the respective places.

File: examples/vision/zero_dce.py (hunk 5 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Changed `SpatialConsistancyLoss` to a subclass of `keras.losses.Loss` and changed the `__call__(original, enhanced)` signature to `call(y_true, y_pred)`.

File: examples/vision/zero_dce.py (hunk 7 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Renamed `_calculate_losses` to `compute_losses`. However, I didn't think that making it a standalone utility function would work since it has a dependency on `self.spatial_constancy_loss`. Instead, I made `get_enhanced_image` into a standalone utility function since it has not dependency on `self`.

File: examples/vision/zero_dce.py (hunk 8 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Removed redundant computation of `output`.

File: examples/vision/zero_dce.py (hunk 2 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py:183]: @fchollet , used a tuple in `tf.reduce_mean`. However, I didn't understand this part, "Pass `axis` as a keyword arg for better readability", can you please share an example?

File: examples/vision/zero_dce.py (hunk 6 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Used `tf.nn`

**Comment by soumik12345** [comment]
Hi @fchollet , I have the following queries:

- I have tried to Include a basic explanation or curve estimation along with relevant references without delving too deep into the mathematics, can you please check if that will be intuitive enough for our readers or not?
- I have some doubts regarding one of your suggestions regarding keyword arg. Can you please clarify this portion?

File: examples/vision/zero_dce.py (hunk 2 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Tried adding axis as a keyword arg, can you please take a look and check if its okay or not?

File: examples/vision/zero_dce.py [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+"""
+### Color Constancy Loss
+
+The Color Constancy Loss is used to correct the potential color deviations in the
+enhanced image and also build the relations among the three adjusted channels.
+"""
+
+
+def color_constancy_loss(x, **kwargs):
+    mean_rgb = tf.reduce_mean(x, kwargs.get("axis", (1, 2)), keepdims=True)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:183]: Let me explain what I meant.

In the statement, `tf.reduce_mean(x, (1, 2))`, the reader doesn't know what `(1, 2)` means. So the statement is not very readable / clear.

For this reason, it's better to use `tf.reduce_mean(x, axis=(1, 2))`. Then the reader knows what is going on: the axes of reduction are being specified. It's more explicit.

In the first case you're passing the argument by position. In the second case you're passing it as a keyword argument (the "keyword" being "axis").

File: examples/vision/zero_dce.py [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+### Exposure loss
+
+To restrain under-/over-exposed regions, the Exposure Control Loss is used. The exposure
+control loss measures the distance between the average intensity value of a local region
+to the well-exposedness level which is set to `0.6`.
+"""
+
+
+def exposure_loss(x, **kwargs):
+    x = tf.reduce_mean(x, kwargs.get("axis", 3), keepdims=True)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:201]: Likewise here

File: examples/vision/zero_dce.py [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+        self.dce_model = build_dce_net()
+
+    def compile(self, learning_rate, **kwargs):
+        super(ZeroDCE, self).compile(**kwargs)
+        self.optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
+        self.spatial_constancy_loss = SpatialConsistancyLoss(reduction="none")
+
+    def call(self, data):
+        dce_net_output = self.dce_model(data)
+        return get_enhanced_image(data, dce_net_output)
```
Comment (fchollet) [inline examples/vision/zero_dce.py:338]: This is part of the forward pass of the model so let's make it a method.

In general:

- Something that intrinsically belongs on the model, like the forward pass or something that uses `self`, should be a method
- Something that is standalone and reusable elsewhere should be a utility (like a loss function or a metric)

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: examples/vision/zero_dce.py (hunk 10 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Thanks for the explanation. I used keyword args for better readability.

File: examples/vision/zero_dce.py (hunk 11 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: I used keyword args for better readability.

File: examples/vision/zero_dce.py (hunk 12 above)
Comment (soumik12345) [reply examples/vision/zero_dce.py]: Made `get_enhanced_image` a method instead of a standalone function.

**Comment by soumik12345** [comment]
Hi @fchollet , thanks for your suggestions and clarifications. I have updated the example accordingly. Can you please review it?

**Review by fchollet — APPROVED** [review]
Awesome! I pushed some copyedits. I believe we're ready to publish -- pull pull the copyedits, then you can add the generated files.

**Comment by soumik12345** [comment]
Hi @fchollet , I added the generated files. Thanks a lot for your reviews.

**Comment by fchollet** [comment]
Thank you for the great contribution! 👍
